### PR TITLE
Four points for enhancing Paddle-TRT

### PR DESCRIPTION
**Pull request by NHZlX**
1. Bind trt input and output to fluid tensors
2. Restruct TRTEngine op: TRTEngine using stream only when executing.
3. When runing in trt mode, do not allocate memory for parameters in fluid.
4. Support loading static model for trt
5. update ci trt version from 4.03 to 4.16.

### CODE REVIEW

File: paddle/fluid/framework/ir/fuse_pass_base.h [hunk 1]
```diff
@@ -24,6 +25,10 @@ namespace ir {
 
 static const char kParamScopeAttr[] = "__param_scope__";
 static const char kFuseStatisAttr[] = "__fuse_statis__";
+// When we use trt or other third_party lib, the parameters are managered by
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/fuse_pass_base.h]: managed

File: paddle/fluid/inference/analysis/ir_passes/tensorrt_subgraph_pass.cc [hunk 2]
```diff
@@ -222,26 +191,82 @@ void TensorRtSubgraphPass::CreateTensorRTOp(framework::ir::Node *node,
...
+  // When in int8 mode and calibration_mode, the program just produce the
+  // calibration table data.
+  bool calibration_mode = (enable_int8 && calibration_data.size() == 0);
+  if (!calibration_mode) {
+    std::copy(params.begin(), params.end(),
+              std::back_inserter(*repetitive_params));
+    std::string trt_engine_serialized_data = GetTrtEngineSerializedData(
+        Get<std::string>("model_opt_cache_dir"), engine_key);
+
+    if (trt_engine_serialized_data.size() == 0) {
```
Comment (Superjomn) [inline paddle/fluid/inference/analysis/ir_passes/tensorrt_subgraph_pass.cc]:
```
if (trt_engine_serialized_data.empty())
```

File: paddle/fluid/inference/tensorrt/test_engine.cc [hunk 3]
```diff
@@ -27,19 +29,29 @@ namespace tensorrt {
...
+    ctx_ = new platform::CUDADeviceContext(platform::CUDAPlace(0));
+
+    engine_ = new TensorRTEngine(10, 1 << 10);
     engine_->InitNetwork();
   }
 
-  void TearDown() override {
-    delete engine_;
-    cudaStreamDestroy(stream_);
+  void TearDown() override { delete engine_; }
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/test_engine.cc]:
```
if (engine_) {
  delete engine_;
  engine_ = nullptr;
}
```

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h [hunk 4]
```diff
@@ -180,129 +148,106 @@ class TensorRTEngineOp : public framework::OperatorBase {
...
+    // Execute the engine.
+    engine->Execute(runtime_batch, &buffers, stream);
     cudaStreamSynchronize(stream);
   }
 
-  void Prepare(const framework::Scope &scope, const platform::Place &dev_place,
-               TensorRTEngine *engine) const {
+  TensorRTEngine *GetEngine(const framework::Scope &scope,
+                            const platform::Place &dev_place) const {
+    if (trt_engine_.get() == nullptr) {
```
Comment (Superjomn) [inline paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]:
```
if (!trt_engine)
```

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h [hunk 5]
```diff
@@ -180,129 +148,106 @@ class TensorRTEngineOp : public framework::OperatorBase {
...
+      trt_engine_.reset(new inference::tensorrt::TensorRTEngine(
+          max_batch_size_, workspace_size_, enable_int8_, calibrator_.get(),
+          boost::get<platform::CUDAPlace>(dev_place).device));
+      if (engine_serialized_data_.size() > 0) {
```
Comment (Superjomn) [inline paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]:
```
if (! engine_serialized_data_.empty())
```

**Review by Superjomn — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h (hunk 5 above)
Comment (NHZlX) [reply paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]: done!

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h (hunk 4 above)
Comment (NHZlX) [reply paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]: done！

File: paddle/fluid/inference/tensorrt/test_engine.cc (hunk 3 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/test_engine.cc]: done！

File: paddle/fluid/inference/analysis/ir_passes/tensorrt_subgraph_pass.cc (hunk 2 above)
Comment (NHZlX) [reply paddle/fluid/inference/analysis/ir_passes/tensorrt_subgraph_pass.cc]: done！

File: paddle/fluid/framework/ir/fuse_pass_base.h (hunk 1 above)
Comment (NHZlX) [reply paddle/fluid/framework/ir/fuse_pass_base.h]: done！

**Review by Superjomn — DISMISSED** [review]
LGTM

**Review by panyx0718 — DISMISSED** [review]
approve const_cast removal

**Review by Superjomn — APPROVED** [review]
LGTM

**Review by panyx0718 — APPROVED** [review]
Approve const_cast removal

**Pull request merged**